{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import pysam\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def index_bam(bam_file):\n",
    "    # Constructing the command as a string to build bam file index\n",
    "    cmd = \"samtools index \" + bam_file\n",
    "    # Running the command using subprocess\n",
    "    subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_average_read_quality(bam_file):\n",
    "     # Open the BAM file for reading ('rb' stands for read binary)\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "    # Initialize an empty list to store read qualities\n",
    "    read_qualities = []\n",
    "    # Iterate over each read in the BAM file\n",
    "    for read in bam:\n",
    "        # Extend the list with the quality scores of the read, excluding the first 10 and last 10 bases\n",
    "        read_qualities.extend(read.query_qualities[10:-10])\n",
    "    # Close the BAM file after reading\n",
    "    bam.close()\n",
    "    # Compute the average of the read qualities using numpy\n",
    "    average_quality = np.mean(read_qualities)\n",
    "    # Return the computed average read quality\n",
    "    return average_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_coverage_at_least_50(bam_file):\n",
    "    # Set the minimum coverage\n",
    "    min_coverage = 50\n",
    "    # Open the BAM file for reading ('rb' stands for read binary)\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "    # Initialize counters for total positions in the genome and those covered by reads\n",
    "    total_positions = 0\n",
    "    covered_positions = 0\n",
    "    # Iterate over each position in the reference genome\n",
    "    for pileupcolumn in bam.pileup():\n",
    "        # Increment the total position counter\n",
    "        total_positions += 1\n",
    "        # If the number of reads covering the position is greater than or equal to the minimum coverage\n",
    "        if pileupcolumn.n >= min_coverage:\n",
    "            # Increment the covered position counter\n",
    "            covered_positions += 1\n",
    "    # Close the BAM file after processing\n",
    "    bam.close()\n",
    "    # Calculate the percentage of positions that meet the minimum coverage requirement\n",
    "    coverage_percentage = (covered_positions / total_positions) * 100\n",
    "    # Return the percentage of positions that meet the minimum coverage requirement\n",
    "    return coverage_percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_average_mapping_quality(bam_file):\n",
    "    # Open the BAM file for reading ('rb' stands for read binary)\n",
    "    bam = pysam.AlignmentFile(bam_file, \"rb\")\n",
    "    # Extract the mapping quality for each read in the BAM file\n",
    "    mapping_qualities = [read.mapping_quality for read in bam]\n",
    "    # Close the BAM file after reading\n",
    "    bam.close()\n",
    "    # Compute the average of the mapping qualities using numpy\n",
    "    average_mapping_quality = np.mean(mapping_qualities)\n",
    "    # Return the computed average mapping quality\n",
    "    return average_mapping_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_average_depth_coverage(bam_file):\n",
    "    # Construct the command as a string\n",
    "    cmd = \"samtools depth \" + bam_file\n",
    "    # Use samtools to compute the depth of the BAM file and capture the output\n",
    "    result = subprocess.run(cmd, capture_output=True, text=True, check=True, shell=True)\n",
    "    # Convert the output directly into a DataFrame\n",
    "    df = pd.DataFrame([line.split('\\t') for line in result.stdout.strip().split('\\n')], columns=['Ref', 'Pos', 'Depth'])\n",
    "    # Convert the 'Depth' column to integer type and compute the mean depth\n",
    "    return df['Depth'].astype(int).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the list of samples\n",
    "samples = [\"04.B1.W14.01\", \"04.M1.W09.02\", \n",
    "           \"05.B1.W14.04\", \"05.M1.W08.03\",\n",
    "           \"27.B1.W13.06\", \"27.M1.W10.07\", \n",
    "           \"30.B1.W11.08\", \"30.M1.W04.09\", \n",
    "           \"38.B1.W10.11\", \"38.M1.W03.10\", \n",
    "           \"39.B1.W11.12\", \"39.M1.W03.13\", \"39.M1.W05.14\", \n",
    "           \"53.B1.W14.17\", \"53.M1.W07.16\", \n",
    "           \"56.B1.W09.22\", \"56.M1.W03.21\", \n",
    "           \"63.B1.W09.29\", \"63.M1.W02.30\", \n",
    "           \"66.B1.W09.25\", \"66.M1.W02.24\"]\n",
    "\n",
    "# Create an empty DataFrame to store the results\n",
    "results_df = pd.DataFrame(columns=['Sample', 'Average read quality', 'Coverage percentage (>50)',\n",
    "                                   'Average mapping quality', 'Average depth coverage'])\n",
    "\n",
    "# Iterate on samples to extract read informations\n",
    "for sample in samples:\n",
    "    bam_file = \"data/\" + sample + \"/\" + sample + \".bam\"\n",
    "    index_bam(bam_file)\n",
    "    average_read_quality = compute_average_read_quality(bam_file)\n",
    "    coverage_at_least_50 = compute_coverage_at_least_50(bam_file)\n",
    "    average_mapping_quality = compute_average_mapping_quality(bam_file)\n",
    "    average_depth_coverage = compute_average_depth_coverage(bam_file)\n",
    "\n",
    "    # Create a new DataFrame with the results for the current sample\n",
    "    sample_df = pd.DataFrame({\n",
    "        'Sample': [sample],\n",
    "        'Average read quality': [round(average_read_quality, 2)],\n",
    "        'Coverage percentage (>50)': [round(coverage_at_least_50, 2)],\n",
    "        'Average mapping quality': [round(average_mapping_quality, 2)],\n",
    "        'Average depth coverage': [round(average_depth_coverage, 2)]})\n",
    "\n",
    "    # Concatenate the new DataFrame with the existing results DataFrame\n",
    "    results_df = pd.concat([results_df, sample_df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Average read quality</th>\n",
       "      <th>Coverage percentage (&gt;50)</th>\n",
       "      <th>Average mapping quality</th>\n",
       "      <th>Average depth coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>04.B1.W14.01</td>\n",
       "      <td>22.02</td>\n",
       "      <td>99.04</td>\n",
       "      <td>57.46</td>\n",
       "      <td>14303.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>04.M1.W09.02</td>\n",
       "      <td>22.12</td>\n",
       "      <td>99.18</td>\n",
       "      <td>56.80</td>\n",
       "      <td>32548.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>05.B1.W14.04</td>\n",
       "      <td>21.94</td>\n",
       "      <td>99.32</td>\n",
       "      <td>58.70</td>\n",
       "      <td>11911.52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>05.M1.W08.03</td>\n",
       "      <td>21.87</td>\n",
       "      <td>99.11</td>\n",
       "      <td>58.54</td>\n",
       "      <td>19751.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27.B1.W13.06</td>\n",
       "      <td>22.34</td>\n",
       "      <td>98.64</td>\n",
       "      <td>55.58</td>\n",
       "      <td>26189.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>27.M1.W10.07</td>\n",
       "      <td>22.16</td>\n",
       "      <td>99.11</td>\n",
       "      <td>56.09</td>\n",
       "      <td>14538.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>30.B1.W11.08</td>\n",
       "      <td>21.97</td>\n",
       "      <td>99.11</td>\n",
       "      <td>57.81</td>\n",
       "      <td>12580.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>30.M1.W04.09</td>\n",
       "      <td>22.04</td>\n",
       "      <td>99.25</td>\n",
       "      <td>58.02</td>\n",
       "      <td>11404.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>38.B1.W10.11</td>\n",
       "      <td>22.25</td>\n",
       "      <td>99.25</td>\n",
       "      <td>56.34</td>\n",
       "      <td>21268.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>38.M1.W03.10</td>\n",
       "      <td>22.26</td>\n",
       "      <td>99.32</td>\n",
       "      <td>56.93</td>\n",
       "      <td>13532.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>39.B1.W11.12</td>\n",
       "      <td>21.89</td>\n",
       "      <td>99.11</td>\n",
       "      <td>56.06</td>\n",
       "      <td>18708.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>39.M1.W03.13</td>\n",
       "      <td>21.99</td>\n",
       "      <td>98.84</td>\n",
       "      <td>56.23</td>\n",
       "      <td>19535.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>39.M1.W05.14</td>\n",
       "      <td>21.99</td>\n",
       "      <td>99.11</td>\n",
       "      <td>56.45</td>\n",
       "      <td>12722.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>53.B1.W14.17</td>\n",
       "      <td>22.28</td>\n",
       "      <td>99.11</td>\n",
       "      <td>56.89</td>\n",
       "      <td>16757.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>53.M1.W07.16</td>\n",
       "      <td>22.34</td>\n",
       "      <td>99.59</td>\n",
       "      <td>56.67</td>\n",
       "      <td>8166.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>56.B1.W09.22</td>\n",
       "      <td>22.03</td>\n",
       "      <td>99.38</td>\n",
       "      <td>57.51</td>\n",
       "      <td>2506.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>56.M1.W03.21</td>\n",
       "      <td>22.01</td>\n",
       "      <td>99.38</td>\n",
       "      <td>57.10</td>\n",
       "      <td>11884.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>63.B1.W09.29</td>\n",
       "      <td>21.95</td>\n",
       "      <td>99.38</td>\n",
       "      <td>59.29</td>\n",
       "      <td>12090.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>63.M1.W02.30</td>\n",
       "      <td>21.94</td>\n",
       "      <td>99.18</td>\n",
       "      <td>59.28</td>\n",
       "      <td>10486.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>66.B1.W09.25</td>\n",
       "      <td>22.31</td>\n",
       "      <td>98.91</td>\n",
       "      <td>56.44</td>\n",
       "      <td>12534.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>66.M1.W02.24</td>\n",
       "      <td>21.96</td>\n",
       "      <td>99.25</td>\n",
       "      <td>56.68</td>\n",
       "      <td>14108.21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample  Average read quality  Coverage percentage (>50)  \\\n",
       "0   04.B1.W14.01                 22.02                      99.04   \n",
       "1   04.M1.W09.02                 22.12                      99.18   \n",
       "2   05.B1.W14.04                 21.94                      99.32   \n",
       "3   05.M1.W08.03                 21.87                      99.11   \n",
       "4   27.B1.W13.06                 22.34                      98.64   \n",
       "5   27.M1.W10.07                 22.16                      99.11   \n",
       "6   30.B1.W11.08                 21.97                      99.11   \n",
       "7   30.M1.W04.09                 22.04                      99.25   \n",
       "8   38.B1.W10.11                 22.25                      99.25   \n",
       "9   38.M1.W03.10                 22.26                      99.32   \n",
       "10  39.B1.W11.12                 21.89                      99.11   \n",
       "11  39.M1.W03.13                 21.99                      98.84   \n",
       "12  39.M1.W05.14                 21.99                      99.11   \n",
       "13  53.B1.W14.17                 22.28                      99.11   \n",
       "14  53.M1.W07.16                 22.34                      99.59   \n",
       "15  56.B1.W09.22                 22.03                      99.38   \n",
       "16  56.M1.W03.21                 22.01                      99.38   \n",
       "17  63.B1.W09.29                 21.95                      99.38   \n",
       "18  63.M1.W02.30                 21.94                      99.18   \n",
       "19  66.B1.W09.25                 22.31                      98.91   \n",
       "20  66.M1.W02.24                 21.96                      99.25   \n",
       "\n",
       "    Average mapping quality  Average depth coverage  \n",
       "0                     57.46                14303.89  \n",
       "1                     56.80                32548.05  \n",
       "2                     58.70                11911.52  \n",
       "3                     58.54                19751.16  \n",
       "4                     55.58                26189.03  \n",
       "5                     56.09                14538.82  \n",
       "6                     57.81                12580.72  \n",
       "7                     58.02                11404.77  \n",
       "8                     56.34                21268.29  \n",
       "9                     56.93                13532.50  \n",
       "10                    56.06                18708.27  \n",
       "11                    56.23                19535.34  \n",
       "12                    56.45                12722.84  \n",
       "13                    56.89                16757.19  \n",
       "14                    56.67                 8166.33  \n",
       "15                    57.51                 2506.24  \n",
       "16                    57.10                11884.31  \n",
       "17                    59.29                12090.93  \n",
       "18                    59.28                10486.46  \n",
       "19                    56.44                12534.15  \n",
       "20                    56.68                14108.21  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the quality of sequencing data\n",
    "results_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
