{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "sys.path.append('utils')\n",
    "import utils\n",
    "import kanalyzer\n",
    "import protein_score\n",
    "import mutation_score\n",
    "import discriminative_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = {\n",
    "        \"UL33\": {\n",
    "        \"k\": 9,\n",
    "        \"taxon_id\": 10359,\n",
    "        \"input_file\": \"data/fasta/UL33_all_nucleotide_sequences_annotated.fasta\",\n",
    "        \"output_folder\": \"data/results\",\n",
    "        \"infos_file\": \"data/UL33_infos.json\",\n",
    "        \"aligner_matrix\": \"BLOSUM62\",\n",
    "        \"aligner_open_gap_score\": -10.0,\n",
    "        \"aligner_extend_gap_score\": -2.0,\n",
    "        \"threshold\": 25\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compile_results(results, threshold_percentage):\n",
    "    \"\"\"\n",
    "    Compile results by aggregating variations and amino acid changes based on a threshold.\n",
    "\n",
    "    Parameters:\n",
    "    - results: Dict containing the analysis results from kanalyzer.\n",
    "    - threshold_percentage: The threshold percentage to include variations.\n",
    "\n",
    "    Returns:\n",
    "    - compiled_results: Dict containing compiled variations and amino acid changes.\n",
    "    \"\"\"\n",
    "    temp_variations = {}\n",
    "\n",
    "    # Collect class counts and temporary variations data\n",
    "    for gene, sequences in results.items():\n",
    "        class_counts = {}\n",
    "        temp_variations.setdefault(gene, {})\n",
    "        \n",
    "        for description, mutations in sequences.items():\n",
    "            class_id = description.split('|')[-1]\n",
    "\n",
    "            class_counts[class_id] = class_counts.get(class_id, 0) + 1\n",
    "\n",
    "            for position_kmer, details in mutations.items():\n",
    "                variation = details['variations']\n",
    "                amino_acid_changes = details['amino_acid_changes']\n",
    "\n",
    "                temp_variations[gene].setdefault(position_kmer, {})\n",
    "                temp_variations[gene][position_kmer].setdefault(variation, {\n",
    "                    'count': 0,\n",
    "                    'amino_acid_changes': set(),\n",
    "                    'class_specific_counts': {}\n",
    "                })\n",
    "\n",
    "                variation_info = temp_variations[gene][position_kmer][variation]\n",
    "                variation_info['count'] += 1\n",
    "                variation_info['amino_acid_changes'].update(amino_acid_changes)\n",
    "                variation_info['class_specific_counts'][class_id] = variation_info['class_specific_counts'].get(class_id, 0) + 1\n",
    "\n",
    "    compiled_results = {}\n",
    "    # Compile results based on the threshold\n",
    "    for gene, positions in temp_variations.items():\n",
    "        compiled_results[gene] = {}\n",
    "        for position_kmer, variations in positions.items():\n",
    "            compiled_results[gene][position_kmer] = {\n",
    "                'variations': set(),\n",
    "                'amino_acid_changes': set()\n",
    "            }\n",
    "            for variation, info in variations.items():\n",
    "                include_variation = False\n",
    "                for class_id, count in info['class_specific_counts'].items():\n",
    "                    total = class_counts[class_id]\n",
    "                    if (count / total) * 100 >= threshold_percentage:\n",
    "                        include_variation = True\n",
    "                        break\n",
    "                if include_variation:\n",
    "                    compiled_results[gene][position_kmer]['variations'].add(variation)\n",
    "                    compiled_results[gene][position_kmer]['amino_acid_changes'].update(info['amino_acid_changes'])\n",
    "    return compiled_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_kss_scores(compiled_results, results, infos):\n",
    "    \"\"\"\n",
    "    Compute k-mer significance scores (KSS) for compiled results.\n",
    "\n",
    "    Parameters:\n",
    "    - compiled_results: Dict containing compiled variations and amino acid changes.\n",
    "    - results: Original results from kanalyzer.\n",
    "    - infos: Dict containing additional information like taxon_id and output_folder.\n",
    "\n",
    "    Returns:\n",
    "    - compiled_results: Updated compiled_results with KSS scores.\n",
    "    \"\"\"\n",
    "    for gene, data in compiled_results.items():\n",
    "        # Update protein score in results\n",
    "        ps = protein_score.get_protein_score(infos['taxon_id'], gene)\n",
    "        for key, details in data.items():\n",
    "            details['protein_score'] = ps\n",
    "\n",
    "        # Evaluate gene-related scores\n",
    "        y = [entry.split('|')[-1] for entry in results[gene].keys()]\n",
    "        categorized_scores_dict = {}\n",
    "        uncategorized_scores_dict = {}\n",
    "\n",
    "        for key, details in data.items():\n",
    "            amino_acid_changes = details['amino_acid_changes']\n",
    "            mutational_scores = mutation_score.get_mutational_scores(\n",
    "                amino_acid_changes, categorize=True\n",
    "            )\n",
    "            mutational_score = max(mutational_scores.values()) if mutational_scores else 0\n",
    "            details['mutational_score'] = mutational_score\n",
    "\n",
    "            X = []\n",
    "            variations = details['variations']\n",
    "            for entry_value in results[gene].values():\n",
    "                variation_presence = [\n",
    "                    1 if variation in entry_value.get(key, {}).get(\"variations\", \"\")\n",
    "                    else 0 for variation in variations\n",
    "                ]\n",
    "                X.append(variation_presence)\n",
    "            \n",
    "            if not X:\n",
    "                continue  # Skip if X is empty\n",
    "\n",
    "            categorized_scores, uncategorized_scores = discriminative_score.get_discriminative_scores(X, y)\n",
    "            categorized_scores_dict[str(key)] = categorized_scores\n",
    "            uncategorized_scores_dict[str(key)] = uncategorized_scores\n",
    "\n",
    "            details['discriminative_score'] = categorized_scores[3]\n",
    "            average_score = (\n",
    "                details['discriminative_score'] +\n",
    "                details['mutational_score'] +\n",
    "                details['protein_score']\n",
    "            ) / 3\n",
    "            details['k_mer_significance_score'] = round(average_score, 1)\n",
    "\n",
    "        # Save categorized and uncategorized scores\n",
    "        output_folder = infos['output_folder']\n",
    "        utils.save_data_as_json(\n",
    "            categorized_scores_dict,\n",
    "            os.path.join(output_folder, f\"{gene}_categorized_scores.json\")\n",
    "        )\n",
    "        utils.save_data_as_json(\n",
    "            uncategorized_scores_dict,\n",
    "            os.path.join(output_folder, f\"{gene}_uncategorized_scores.json\")\n",
    "        )\n",
    "    return compiled_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Processing: UL33\n",
      "Results saved to data/results\\UL33_results.json\n",
      "Compiled results saved to data/results\\UL33_compiled_results.json\n"
     ]
    }
   ],
   "source": [
    "# Main processing loop\n",
    "for dataset, infos in datasets.items():\n",
    "    print(f\"\\nProcessing: {dataset}\")\n",
    "    # Load configuration and input files\n",
    "    threshold = infos['threshold']\n",
    "    input_file = infos['input_file']\n",
    "    output_folder = infos['output_folder']\n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "    # Load all sequences once to avoid opening the file multiple times\n",
    "    with open(input_file, \"r\") as file:\n",
    "        records = list(SeqIO.parse(file, \"fasta\"))\n",
    "\n",
    "    # Analyze records using kanalyzer\n",
    "    results = kanalyzer.analyze_records(records, infos)\n",
    "\n",
    "    # Save initial results\n",
    "    results_file_path = os.path.join(output_folder, f\"{dataset}_results.json\")\n",
    "    utils.save_data_as_json(results, results_file_path)\n",
    "    print(f\"Results saved to {results_file_path}\")\n",
    "\n",
    "    # Compile results and save\n",
    "    compiled_results = compile_results(results, threshold)\n",
    "    compiled_results = compute_kss_scores(compiled_results, results, infos)\n",
    "    compiled_file_path = os.path.join(output_folder, f\"{dataset}_compiled_results.json\")\n",
    "    utils.save_data_as_json(compiled_results, compiled_file_path)\n",
    "    print(f\"Compiled results saved to {compiled_file_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  },
  "vscode": {
   "interpreter": {
    "hash": "84ef65aefc9850a1f4201c08de368b8c2cca283fb4b9eac39969d668c5d9d9f0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
